battle.go

```diff
@@ -1,115 +1,212 @@
+	return nil
+}
+
+// Set which agents have which allies. Not to be confused with `Party`.
+func (b *Battle) SetTeams(t [][]int) error {
+	// TODO: validate
+	b.teams = t
+	return nil
+}
+
+type BattleState int
 
-const (
-	PLAYER1 Player = iota
-	PLAYER2
-)
 const (
 	BEFORE_START BattleState = iota
-	BATTLE_START
-	BEFORE_PLAYER1_TURN
-	AFTER_PLAYER1_TURN
-	BEFORE_PLAYER2_TURN
-	AFTER_PLAYER2_TURN
+	BATTLE_IN_PROGRESS
 	BATTLE_END
 )
 
-// Creates a new battle instance that agents can connect to
-func NewBattle() *Battle {
-	return &Battle{State: BEFORE_START}
+// Adds Agent(s) to the battle.
+func (b *Battle) AddAgent(a ...*Agent) {
+	b.Agents = append(b.Agents, a...)
 }
 
-// Updates the battle state and dispatches calls to custom event hooks
-func (b *Battle) Dispatch(state BattleState) {
-	b.State = state
-	for _, player := range b.Players {
-		if handler, ok := player.EventHooks[state]; ok {
-			handler()
-		}
-	}
+// Adds Parties to the battle.
+func (b *Battle) AddParty(p ...*Party) {
+	b.Parties = append(b.Parties, p...)
 }
 
-// Adds player(s) to the battle
-// Fails if two players are already connected
-func (b *Battle) AddAgents(a ...*Agent) {
-	if len(b.Players)+len(a) > 2 {
-		log.Panicf("battle cannot have more than two agents connected")
+func (b *Battle) Start() error {
+	// TODO: validate the battle, return error if invalid
+
+	// Initiate the battle! Send out the first pokemon in the parties.
+	b.State = BATTLE_IN_PROGRESS
+	for _, p := range b.Parties {
+		p.SetActive(0)
 	}
-	b.Players = append(b.Players, a...)
+	return nil
 }
 
-// Removes a player by their ID from the battle
-func (b *Battle) RemoveAgent(agentID int64) {
-	players := make([]*Agent, 0)
-	for _, p := range b.Players {
-		if p.ID != agentID {
-			players = append(players, p)
+// Simulates a single round of the battle.
+func (b *Battle) SimulateRound() {
+	// get the currently active pokemon
+	active := []activePokemon{}
+	for a := range b.Agents {
+		party := b.Parties[b.agentParties[a]]
+		for _, idx := range party.GetActive() {
+			ap := activePokemon{
+				AgentIdx:   a,
+				PokemonIdx: idx,
+			}
+			active = append(active, ap)
 		}
 	}
-	if len(players) == len(b.Players) {
-		log.Panicf("no agent with ID '%v' found", agentID)
+
+	context := battleContext{
+		Context:       *b,
+		ActivePokemon: active,
+	}
+
+	// get turns from agents
+	// map of active pokemon indexes to their turns
+	turns := map[int]Turn{}
+	for i, ap := range active {
+		context.SetSelf(i)
+		agent := *b.Agents[ap.AgentIdx]
+		turn := agent.Act(&context)
+		turns[i] = turn
 	}
-	b.Players = players
+
+	turnOrder := []int{}
+	// TODO: determine the correct order based on priority and other factors
+	for i := range active {
+		turnOrder = append(turnOrder, i)
+	}
+
+	for _, apIdx := range turnOrder {
+		switch t := turns[apIdx].(type) {
+		case FightTurn:
+			fmt.Printf("TODO: Implement fight %v\n", t)
+		default:
+			panic("Unknown turn")
+		}
+	}
+}
```

Comment:
Can't we just calculate priority based on the context and then sort the turns in order of priority?

Reply:
That is precisely what should happen here. I left it out intentionally for somebody else to implement.

`turns` maps active pokemon to their planned actions. We can't sort the turns directly because they depend on the order of the active pokemon.